cdap-security/src/test/java/io/cdap/cdap/security/authorization/InMemoryAccessController.java

```diff
@@ -0,0 +1,320 @@
+import io.cdap.cdap.proto.security.GrantedPermission;
+import io.cdap.cdap.proto.security.Permission;
+import io.cdap.cdap.proto.security.Principal;
+import io.cdap.cdap.proto.security.Role;
+import io.cdap.cdap.security.spi.authorization.AccessController;
+import io.cdap.cdap.security.spi.authorization.AlreadyExistsException;
+import io.cdap.cdap.security.spi.authorization.AuthorizationContext;
+import io.cdap.cdap.security.spi.authorization.NotFoundException;
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * In-memory implementation of {@link AccessController}.
+ */
+public class InMemoryAccessController implements AccessController {
+
+  private final ConcurrentMap<Authorizable, ConcurrentMap<Principal, Set<Permission>>> privileges =
+    new ConcurrentHashMap<>();
+  private final ConcurrentMap<Role, Set<Principal>> roleToPrincipals = new ConcurrentHashMap<>();
+  private final Set<Principal> superUsers = new HashSet<>();
+  // Bypass enforcement for tests that want to simulate every user as a super user
+  private final Principal allSuperUsers = new Principal("*", Principal.PrincipalType.USER);
+
+  @Override
+  public void destroy() {
+
+  }
+
+  @Override
+  public void initialize(AuthorizationContext context) {
+    Properties properties = context.getExtensionProperties();
+    if (properties.containsKey("superusers")) {
+      for (String superuser : Splitter.on(",").trimResults().omitEmptyStrings()
+        .split(properties.getProperty("superusers"))) {
+        superUsers.add(new Principal(superuser, Principal.PrincipalType.USER));
+      }
+    }
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws UnauthorizedException {
+    // super users do not have any enforcement
+    if (superUsers.contains(principal) || superUsers.contains(allSuperUsers)) {
+      return;
+    }
+    // permissions allowed for this principal
+    Set<? extends Permission> allowed = getPermissions(entity, principal);
+    if (allowed.containsAll(permissions)) {
+      return;
+    }
+    Set<Permission> allowedForRoles = new HashSet<>();
+    // permissions allowed for any of the roles to which this principal belongs if its not a role
+    if (principal.getType() != Principal.PrincipalType.ROLE) {
+      for (Role role : getRoles(principal)) {
+        allowedForRoles.addAll(getPermissions(entity, role));
+      }
+    }
+    if (!allowedForRoles.containsAll(permissions)) {
+      throw new UnauthorizedException(principal, Sets.difference(permissions, allowed), entity);
+    }
+  }
+
+  @Override
+  public Set<? extends EntityId> isVisible(Set<? extends EntityId> entityIds, Principal principal) {
```

Comment:
non-actionable: FYI this is an example of what I mean when I said that the isVisible parent checking implementation is delegated to the extension; if we plan on keeping this assumption, we should probably codify the assumption within the new layer instead of delegating the behavior to the extension.

Reply:
We can get rid of isVisible in the subsequent PR as a P2

Reply:
Acknowledged, SGTM